text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+isn't that different from
+```rust
+foo.enum#discriminant
+```
+
+And of course you can always make a function to give it a shorter name -- or write
+a proc macro to generate that function -- if you so wish.
+
+## Why just `pub(in self)`?
+
+The primary use case that led to this RFC is using it in `derive` macros, where
+`pub(in self)` is entirely sufficient.
+
+And by being only private, it avoids forcing any semver promises on library authors.
+
+Today, as a library author, you can reorder the variants in an enum should you so wish,
+or in a `#[non_exhausive]` enum add new ones in the middle.  There's no way for
+the users of your library to care about the order in which you defined the variants
+(unless you make other documented promises) -- especially if you never `derive(PartialOrd)`.
+
+Any library author who wishes to provide discriminant stability can always write
+a function to expose those discriminants, trivially implemented using this feature.
+
+## Why expose it via `.`?
+
+I like it behaving kinda like a field.  For example, having auto-deref like a field
+means you don't need to worry about whether you actually have a `&&Enum` in a `filter`
+or you actually have a `Box<Enum>` or whatever.
+
+Of course, if the `enum` is `repr(C)`, then the discriminant [is a field][RFC2195]
+in the guaranteed FFI layout, so thinking of it kinda like a field isn't too weird.
+
+There has also been talk of *compressed* or *move-only* fields where getting the
+address is disallowed so that Rust can run arbitrary logic whenever they're accessed
+and thus have the freedom to do more layout optimizations than are otherwise possible.
+Should we have something like that, then it's again not unreasonable to think of it
+as a field that sometimes has particularly fancy layout optimization.
+
+[RFC2195]: https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html
+
+## What about if it was a magic method instead?
+
+It could be.  But it would still need to be something that doesn't cause name
+resolution failures for other methods that people might already have written.
+
+So I don't think that the extra `()` on it would really improve things.
+
+## Why not allow writing to the discriminant?
+
+The semantics for that get really complicated, especially for `enum`s in `repr(Rust)`
+that don't have a guaranteed layout, and even more so those that get layout-optimized.
+
+Maybe one day it could be allowed, but for now this RFC sticks only things that
+can be allowed in safe code without worries.
+
+## Couldn't this be a magic macro?
+
+Sure, it could, like `offset_of!`.
+
+I don't think `enum_discriminant!(foo)` is really better than `foo.enum#discriminant`, though.
+
+It doesn't deal in tokens, and there's no special logic to apply to the scope in which
+the argument is computed.
+
+It works on a value or place, not on anything dealing tokens, nor does it affect a scope.
+
+## Why not do *\<more complex feature\>*?
+
+Privacy is the problem.
```

Comment:
How does this proposal fit into the broader SemVer considerations on enum discriminants? Do we currently consider enum discriminants to be under the scope of SemVer? Note, it's _already_ the case that for for fieldless enums without tuple-like variants, `as` casting provides a public mechanism for reflecting the discriminant.

I think the space of options is something like:
- We document that discriminant *never* has SemVer obligations.
- We document that the discriminants defined in edition 202X and up do not have SemVer obligations, and disallow `as` casting of enums defined in greater editions.
- We document that discriminants currently visible *only* by `as`-casting have SemVer obligations, and discriminants not visible by `as` casting do not have SemVer obligations.
- We document that all discriminants have SemVer obligations (in which case the privacy concern here is moot).